libcontainer/container_linux.go

```diff
@@ -1176,7 +1240,9 @@ func (c *linuxContainer) makeCriuRestoreMountpoints(m *configs.Mount) error {
 	case "bind":
 		// The prepareBindMount() function checks if source
 		// exists. So it cannot be used for other filesystem types.
-		if err := prepareBindMount(m, c.config.Rootfs); err != nil {
+		// TODO: pass something else than nil? Not sure if criu is
+		// impacted by issue #2484
```

Comment:
Nope, neither did Alban. To be honest, I never used CRIU before. If you can provide some basic guidelines, it is very much appreciated :)

Reply:
We can add a c/r version of the inaccessible paths test to make sure that we can restore containers with such paths (and if the test fails in CI we know it doesn't work). On paper `runc checkpoint` and `runc restore` should "just work" but it's probably simplest to just try to write an integration test and then we get to double-check it doesn't break in CI as well.

I expect that CRIU has also worked around this issue already but it's possible they haven't.

Reply:
I tried to create a bats test, but it does not work. I've pushed it to https://github.com/kinvolk/runc/tree/rata/users-2484-criu, so you can have a look just in case. I'm not familiar with CRIU and some help will be nice :)

The error that I see running locally is this (attached at the end of this message), it is "Operation not permitted" so I guess there is something we don't have permission to do in the userns, but just a wild guess.

I tried several things to make this work, or even investigated adapting the test "@test "checkpoint and restore", but they also fail (I think it can be made to work and haven't tried enouch, but in any case that didn't help to understand why a checkpoint for this fails).


The error:
```
$ sudo PATH="$PWD:/usr/sbin/:$PATH" bats -t tests/integration/checkpoint.bats -f inaccessible
1..1
not ok 1 checkpoint and restore with inaccessible bind mounts
# (in test file tests/integration/checkpoint.bats, line 475)
#   `[ "$status" -eq 0 ]' failed
# runc spec (status=0):
# 
# runc run -d --console-socket /tmp/bats-run-1544081/runc.VHq1Zy/tty/sock test_busybox (status=0):
# 
# runc state test_busybox (status=0):
# {
#   "ociVersion": "1.0.2-dev",
#   "id": "test_busybox",
#   "pid": 1544189,
#   "status": "running",
#   "bundle": "/tmp/bats-run-1544081/runc.VHq1Zy/bundle",
#   "rootfs": "/tmp/bats-run-1544081/runc.VHq1Zy/bundle/rootfs",
#   "created": "2021-09-22T14:06:38.392579137Z",
#   "owner": ""
# }
# runc --criu /usr/sbin//criu checkpoint --work-path ./work-dir test_busybox (status=0):
# 
# runc state test_busybox (status=1):
# time="2021-09-22T16:06:38+02:00" level=error msg="container does not exist"
# runc --criu /usr/sbin//criu restore -d --work-path ./work-dir --console-socket /tmp/bats-run-1544081/runc.VHq1Zy/tty/sock test_busybox (status=1):
# time="2021-09-22T16:06:38+02:00" level=error msg="criu failed: type NOTIFY errno 0\nlog file: work-dir/restore.log"
# (00.051421)      1: mnt: 	Bind /tmp/.criu.mntns.I81ArM/13-0000000000/dev/pts/0 to /tmp/.criu.mntns.I81ArM/13-0000000000/dev/console
# (00.051438)      1: mnt: 1340:/tmp/.criu.mntns.I81ArM/13-0000000000/dev/console private 1 shared 0 slave 0
# (00.051451)      1: mnt: 	Mounting sysfs @/tmp/.criu.mntns.I81ArM/13-0000000000/sys (0)
# (00.054372)      1: mnt: 1407:/tmp/.criu.mntns.I81ArM/13-0000000000/sys private 1 shared 0 slave 0
# (00.054418)      1: mnt: 	Mounting cgroup2 @/tmp/.criu.mntns.I81ArM/13-0000000000/sys/fs/cgroup (0)
# (00.054489)      1: Error (criu/mount.c:2012): mnt: Unable to mount cgroup /tmp/.criu.mntns.I81ArM/13-0000000000/sys/fs/cgroup (id=1408): Operation not permitted
# (00.054518)      1: Error (criu/mount.c:2077): mnt: Can't mount at /tmp/.criu.mntns.I81ArM/13-0000000000/sys/fs/cgroup: Operation not permitted
# (00.054526)      1: mnt: Start with 0:/tmp/.criu.mntns.I81ArM
# (00.056869) uns: calling exit_usernsd (-1, 1)
# (00.056950) uns: daemon calls 0x55de0d15d110 (1544258, -1, 1)
# (00.056966) uns: `- daemon exits w/ 0
# (00.058166) uns: daemon stopped
# (00.058199) Error (criu/cr-restore.c:2397): Restoring FAILED.
```

Reply:
Note that in the CI the error is permission denied (should have pasted the CI output, I think I have other issues in my local setup for CRIU). Pasted the error here: https://github.com/opencontainers/runc/pull/2576#issuecomment-926634838